flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -3283,13 +3283,47 @@ IntrinsicLibrary::genLbound(mlir::Type resultType,
 fir::ExtendedValue
 IntrinsicLibrary::genUbound(mlir::Type resultType,
                             llvm::ArrayRef<fir::ExtendedValue> args) {
-  assert(args.size() == 3);
-  mlir::Value extent = fir::getBase(genSize(resultType, args));
-  mlir::Value lbound = fir::getBase(genLbound(resultType, args));
+  assert(args.size() == 3 || args.size() == 2);
```

Comment:
I see that the `array` argument was lowered `asAddr` in the table, it should be set to `asInquired` to avoid needless copies of non contiguous arguments.

Reply:
I'm not sure I understand your suggestion.  The current settings for UBOUND are:
```
    {"ubound",
     &I::genUbound,
     {{{"array", asBox}, {"dim", asValue}, {"kind", asValue}}},
     /*isElemental=*/false},
```
If I change the `asBox` to `asInquired`, I get validation errors from the generated MLIR:
```
error: loc("./ub2.f90":3:12): 'fir.convert' op invalid type conversion
FATAL: verification of lowering to FIR failedmodule attributes {fir.defaultkind = "a1c4d8i4l4r4", fir.kindmap = "", llvm.target_triple = "x86_64-unknown-linux-gnu"}  {
```

Here's some sample code that I think is causing the problem:
```
  %14 = fir.convert %0 : (!fir.ref<!fir.box<!fir.heap<!fir.array<?xi32>>>>) -> !fir.ref<!fir.box<none>>
  %15 = fir.convert %9 : (!fir.heap<!fir.array<?x?xi32>>) -> !fir.box<none>
  %16 = fir.convert %c4 : (index) -> i32
  %17 = fir.convert %13 : (!fir.ref<!fir.char<1,10>>) -> !fir.ref<i8>
  %18 = fir.call @_FortranAUbound(%14, %15, %16, %17, %c3_i32_2) : (!fir.ref<!fir.box<none>>, !fir.box<none>, i32, !fir.ref<i8>, i32) -> none
```
I think that the assignment to `%15` is causing the validation error.

Reply:
Never mind, `asBox` is correct here, I thought it was `asAddr` and gave bad advice here.